lib/ssh_scan/crypto.rb

```diff
@@ -5,21 +5,37 @@ module Crypto
     class PublicKey
       def initialize(key)
         @key = key
-        @supported = check_supported
+        @rsa_supported = check_rsa_supported
+        @dsa_supported = check_dsa_supported
         if @key.is_a?(OpenSSL::PKey::RSA)
           @data_string = OpenSSL::ASN1::Sequence([
            OpenSSL::ASN1::Integer.new(@key.public_key.n),
            OpenSSL::ASN1::Integer.new(@key.public_key.e)
           ])
         end
+        if @key.is_a?(OpenSSL::PKey::DSA)
+          @data_string = OpenSSL::ASN1::Sequence([
+           OpenSSL::ASN1::Integer.new(@key.public_key.p),
+           OpenSSL::ASN1::Integer.new(@key.public_key.q),
+           OpenSSL::ASN1::Integer.new(@key.public_key.g)
+          ])
+        end
       end
 
-      def check_supported
+      def check_rsa_supported
```

Comment:
In a case of a high latency action, you can store the result in an instance variable on the class and allow the method to access it, but in this case we're just checking that the key exists and that it's of a specific class, so this should be reasonably performant.